### PR TITLE
Groups Component Endpoints

### LINKED ISSUE #81: Fix Hidden Groups Permissions Check
User's should be able to get hidden groups that they are members of but the `bp_moderate` check is overwriting that.

### PR DESCRIPTION
**Pull request by renatonascalves**
Completing the groups endpoints.

fixes #81.

### CODE REVIEW

File: includes/bp-groups/classes/class-bp-rest-groups-endpoint.php [hunk 1]
```diff
@@ -297,32 +557,96 @@ public function prepare_item_for_response( $item, $request, $is_raw = false ) {
...
+		// Group Creator ID.
+		if ( ! empty( $schema['properties']['creator_id'] ) && isset( $request['user_id'] ) ) {
+			$prepared_group->creator_id = (int) $request['user_id'];
+		} else {
+			$prepared_group->creator_id = get_current_user_id();
+		}
+
+		// Group Slug.
+		if ( ! empty( $schema['properties']['slug'] ) && isset( $request['slug'] ) ) {
+			$prepared_group->slug = $request['slug'];
```
Comment (paulgibbs) [inline includes/bp-groups/classes/class-bp-rest-groups-endpoint.php]: Where are these `$request` value sanitised? The `sanitize_callback` argument says it is one here, but I don't see any input sanitisation for these string values. Is any needed?

File: includes/bp-groups/classes/class-bp-rest-groups-endpoint.php [hunk 2]
```diff
@@ -179,36 +194,279 @@ public function get_item( $request ) {
...
+			return new WP_Error( 'rest_group_invalid_id',
+				__( 'Invalid group id.', 'buddypress' ),
+				array(
+					'status' => 404,
+				)
+			);
+		}
+
+		// If group author does not match logged_in user, block access.
+		if ( get_current_user_id() !== $group->creator_id || ! bp_current_user_can( 'bp_moderate' ) ) {
```
Comment (paulgibbs) [inline includes/bp-groups/classes/class-bp-rest-groups-endpoint.php]: Does core check `creator_id` directly or is there a helper function somewhere?

Comment (renatonascalves) [reply includes/bp-groups/classes/class-bp-rest-groups-endpoint.php]: Actually, no. Most of the time, the delete and update actins are done in the back-end, checking `bp_current_user_can( 'bp_moderate' )`.

I do think something along the lines of `bp_user_can_update_groups` or `bp_user_can_delete_groups` should exist. But admins and the group create should have access to update and delete it.

File: includes/bp-groups/classes/class-bp-rest-groups-endpoint.php [hunk 3]
```diff
@@ -179,36 +194,279 @@ public function get_item( $request ) {
...
+
+	/**
+	 * Update a group.
+	 *
+	 * @since 0.1.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error
+	 */
+	public function update_item( $request ) {
```
[the remark this answers is not in the document]
Comment (renatonascalves) [reply includes/bp-groups/classes/class-bp-rest-groups-endpoint.php]: I updated to reflect this question and added unit tests to confirm this. :)

File: includes/bp-groups/classes/class-bp-rest-groups-endpoint.php (hunk 1 above)
Comment (renatonascalves) [reply includes/bp-groups/classes/class-bp-rest-groups-endpoint.php]: I updated to reflect this question. Indeed, sanitization of some fields were missing.

Comment (renatonascalves) [reply includes/bp-groups/classes/class-bp-rest-groups-endpoint.php]: They are done here! But the values were missing: https://core.trac.wordpress.org/browser/tags/4.9/src/wp-includes/rest-api/class-wp-rest-request.php#L763

**Comment by renatonascalves** [comment]
@paulgibbs Let me know if you have any more suggestions. :)

**Pull request merged**